src/Lean/Elab/Tactic/BVDecide/LRAT/Internal/Formula/RupAddSound.lean

```diff
@@ -0,0 +1,834 @@
+    · apply Or.inr
+      rw [i'_eq_i] at i_true_in_c
+      apply And.intro i_true_in_c
+      simp only [addAssignment, ← b_eq_false, addNegAssignment, ite_false] at h2
+      split at h2
+      · next heq =>
+        have hasPosAssignment_fi : hasAssignment true (f.assignments[i.1]'i_in_bounds) := by
+          simp only [hasAssignment, hasPosAssignment, ite_true, heq]
+        have p_entails_i := (assignmentsInvariant_of_strongAssignmentsInvariant f f_readyForRupAdd.2).2 i true hasPosAssignment_fi p pf
+        simp only [(· ⊨ ·)] at p_entails_i
+        exact p_entails_i
+      · simp only at h2
+      · next heq =>
+        exfalso
+        rw [heq] at h3
+        exact h3 (has_of_both b)
+      · simp only at h2
+  · exfalso
+    have i_true_in_insertUnit_fold : (i, true) ∈ (List.foldl insertUnit (f.rupUnits, f.assignments, false) (negate c)).1.data := by
+      have i_rw : i = ⟨i.1, i.2⟩ := rfl
+      rw [i_rw, ← h1]
+      apply List.get_mem
+    have i_false_in_insertUnit_fold : (i, false) ∈ (List.foldl insertUnit (f.rupUnits, f.assignments, false) (negate c)).1.data := by
+      have i_rw : i = ⟨i.1, i.2⟩ := rfl
+      rw [i_rw, ← h2]
+      apply List.get_mem
+    simp only [f_readyForRupAdd.1, negate, Literal.negate] at i_true_in_insertUnit_fold i_false_in_insertUnit_fold
+    have i_true_in_insertUnit_fold :=
+      mem_insertUnit_fold_units #[] f.assignments false (c.clause.map Literal.negate) (i, true) i_true_in_insertUnit_fold
+    have i_false_in_insertUnit_fold :=
+      mem_insertUnit_fold_units #[] f.assignments false (c.clause.map Literal.negate) (i, false) i_false_in_insertUnit_fold
+    simp only [Literal.negate, List.mem_map, Prod.mk.injEq, Bool.not_eq_true', Prod.exists,
+      exists_eq_right_right, exists_eq_right, Array.data_toArray, List.find?, List.not_mem_nil, or_false,
+      Bool.not_eq_false'] at i_true_in_insertUnit_fold i_false_in_insertUnit_fold
+    have c_not_tautology := Clause.not_tautology c (i, true)
+    simp only [Clause.toList, (· ⊨ ·)] at c_not_tautology
+    rw [DefaultClause.toList] at c_not_tautology
+    rcases c_not_tautology with i_true_not_in_c | i_false_not_in_c
+    · exact i_true_not_in_c i_false_in_insertUnit_fold
+    · exact i_false_not_in_c i_true_in_insertUnit_fold
+
+theorem insertRup_entails_safe_insert {n : Nat} (f : DefaultFormula n) (f_readyForRupAdd : ReadyForRupAdd f)
+    (c : DefaultClause n) :
+    (insertRupUnits f (negate c)).2 = true → Limplies (PosFin n) f (f.insert c) := by
+  intro h p pf
+  simp only [formulaEntails_def, List.all_eq_true, decide_eq_true_eq]
+  intro c' c'_in_fc
+  rw [insert_iff] at c'_in_fc
+  rcases c'_in_fc with c'_eq_c | c'_in_f
+  · rw [c'_eq_c]
+    exact insertRup_entails_hsat f f_readyForRupAdd c p pf h
+  · simp only [formulaEntails_def, List.all_eq_true, decide_eq_true_eq] at pf
+    exact pf c' c'_in_f
+
+theorem insertRupUnits_preserves_AssignmentsInvariant {n : Nat} (f : DefaultFormula n) (f_readyForRupAdd : ReadyForRupAdd f)
```

Comment:
```suggestion
theorem insertRupUnits_preserves_assignmentsInvariant {n : Nat} (f : DefaultFormula n) (f_readyForRupAdd : ReadyForRupAdd f)
```
(or, if you want to go by the naming convention, `assignmentInvariant_insertRupUnits`).